File: interpreter.js
const machineInstructions = require('./instructions');

module.exports = function(memory) {
  // let { heap, registers } = memory;
  function step() {
    const instruction = 
      machineInstructions[memory.heap[memory.inPtr]];
    if (instruction) {
      const params = [ memory ];
      for (let i = 0; i < instruction.paramaterCount; i++) {
        params.push(memory.heap[++memory.inPtr]);
      }
      let traceParams = [instruction.instruction];
      traceParams = traceParams.concat(params);
      // trace.apply({}, traceParams)
      instruction.action.apply(machineInstructions, params);
    }
    memory.inPtr++;
  }

  return {
    step
  };
};

const fs = require('fs');

function trace(instruction, memory, ...rest) {
  fs.appendFileSync('./log.txt', 
    `${ memory.inPtr }: ${ instruction } ${ rest.join(' ') }\n`
  );
}

File: cs/synsharp/Interpreter.cs
using System;
using System.Collections.Generic;

namespace synsharp {
  public class Interpreter {
    private State state;

    private List<Instruction> Instructions;

    public Interpreter(State state) {
      this.state = state;
      this.Instructions = Instruction.GetInstructions();
    }

    public void Step() {
      var instruction = Instructions[state.Heap[state.InPtr]];
      var args = new ushort[instruction.ArgCount];
      if (instruction != null) {
        for (var i = 0; i < instruction.ArgCount; i++) {
          args[i] = state.Heap[++state.InPtr];
        }
        // this.Log(state, instruction, args);
        instruction.Action(state, args);
      }
      state.InPtr++;
    }

    private void Log(State state, Instruction instruction, ushort[] args) {
      using (var stream = System.IO.File.AppendText("./log.txt")) {
        stream.Write(String.Format("{0}: {1} {2}\n", state.InPtr, instruction.OpCode, String.Join(" ", args)));
      }
    }
  }
}